### ISSUE TITLE
Problem with uniform sampling

### ISSUE THREAD

**Issue by yechenw** [fangchangma/sparse-to-dense.pytorch#15]
Hi. I found a discrepancy between the implementation of uniform sampling in dense_to_sparse.py between the method discussed in the paper. Basically, the paper suggests that the Bernoulli probability will be applied to only valid depth pixels from ground truth, whereas the code samples from all pixels. For this reason, the input contains some invalid depth inputs (the ones that are missing in ground truth, with values of zero in the input depth matrix), which is meaningless during training. I've changed the code and apparently it yields better result compared to the original implementation. Here's the modification:

Original:
<img width="657" alt="screen shot 2018-09-07 at 12 02 52 pm" src="https://user-images.githubusercontent.com/22457731/45238229-0431a100-b296-11e8-9595-560fd3137e45.png">

Modified:
<img width="519" alt="screen shot 2018-09-07 at 1 38 53 pm" src="https://user-images.githubusercontent.com/22457731/45242151-db180d00-b2a3-11e8-9090-e2ca3bd1e7cf.png">

It might not be the most efficient way of implementation, but I think the overall idea is correct. Could you please take a look at this? Thanks

**Comment by fangchangma** [comment]
@yechenw Thanks for point this out. I think you are right and the current implementation does count zero pixels in. I wonder, how much of an improvement are you seeing with your new implementation?

There are probably better implementations than iterating over all pixels. However, if you create a pull request I'm happy to merge it.

**Comment by yechenw** [comment]
Thanks for the reply. And yea I think there will definitely be a more efficient way of implementing this using scientific libraries but I'm not exactly sure how to do that. Was just trying the most naive way and see how it does. For the improvement, I think it will definitely be much better as by doing this every depth input to the network would be meaningful, since on the other hand it is possible for the original implementation to get an all-zeros depth input for a image if one gets unlucky. I'm not able to provide the whole result right now as the training is still taking the progress. Currently it is on the fourth epoch and delta1 is about 95% with 500 sparse depth samples, which is higher than 93.5% specified in the paper.

**Comment by fangchangma** [comment]
@yechenw Any update on the final accuracy? Thanks.

**Comment by yechenw** [comment]
Sorry for the late reply. Here's the content of 'best.txt' after training for 15 epochs:

epoch=11
mse=11.415
rmse=3.199
absel=0.054
lg10=nan
mae=1.073
delta1=0.949
t_gpu=0.0094

**Comment by yechenw** [comment]
Hi, any followups?

**Comment by fangchangma** [comment]
Thanks for the update. As mentioned earlier, if you create a pull request, I'm happy to test it out and merge it into master.

**Comment by fangchangma** [comment]
@yechenw Hi. Do you mind sharing the exact command you used?

**Comment by yechenw** [comment]
Hi. It is

--data kitti -m rgbd -s 500 -d upproj

**Comment by fangchangma** [comment]
Hi @yechenw. I merged #16 into the development branch for now.
